pykeops/numpy/utils.py

```diff
@@ -119,6 +133,81 @@ def array(x, dtype=default_dtype, device=None):
     def device(x):
         return "cpu"
 
+    @staticmethod
+    def distance_function(metric):
+        def euclidean(x, y):
+            return ((x - y) ** 2).sum(-1)
+
+        def manhattan(x, y):
+            return ((x - y).abs()).sum(-1)
+
+        def angular(x, y):
+            return -(x | y)
+
+        def angular_full(x, y):
+            return angular(x, y) / ((angular(x, x) * angular(y, y)).sqrt())
+
+        if metric == "euclidean":
+            return euclidean
+        elif metric == "manhattan":
+            return manhattan
+        elif metric == "angular":
+            return angular
+        elif metric == "angular_full":
+            return angular_full
+        elif metric == "hyperbolic":
+            raise ValueError(
+                "Hyperbolic not supported for numpy, please use torch version with approximation"
+            )
+        else:
+            raise ValueError("Unknown metric")
+
+    @staticmethod
+    def sort(x):
+        perm = np.argsort(x)
+        return x[perm], perm
+
+    @staticmethod
+    def unsqueeze(x, n):
+        return np.expand_dims(x, n)
+
+    @staticmethod
+    def arange(n, device="cpu"):
+        return np.arange(n)
+
+    @staticmethod
+    def repeat(x, n):
+        return np.repeat(x, n)
+
+    @staticmethod
+    def to(x, device):
+        return x
+
+    @staticmethod
+    def index_select(input, dim, index):
+        return np.take(input, index, axis=dim)
+
+    @staticmethod
+    def kmeans(x, distance=None, K=10, Niter=15, device="CPU", approx=False, n=0):
+        if distance is None:
+            distance = numpytools.distance_function("euclidean")
+        if approx:
+            raise ValueError("Approx not supported on numpy version")
+        from pykeops.numpy import LazyTensor
+
+        N, D = x.shape
+        c = np.copy(x[:K, :])
+        x_i = LazyTensor(x[:, None, :])
+        for i in range(Niter):
+            c_j = LazyTensor(c[None, :, :])
+            D_ij = distance(x_i, c_j)
+            D_ij.backend = device
+            cl = D_ij.argmin(axis=1).astype(int).reshape(N)
+            Ncl = np.bincount(cl).astype(dtype="float32")
+            for d in range(D):
+                c[:, d] = np.bincount(cl, weights=x[:, d]) / Ncl
```

Comment:
This is probably inefficient: we should investigate the best way of implementing this operation with NumPy.

Reply:
as discussed